docs/buffer-metatable.md

```diff
@@ -0,0 +1,147 @@
+This is similar to function inlining functionality we have where an inlined function will no longer respect its environment.
+
+The freezing is performed to provide additional guarantees for field access and other metamethod evaluation to unlock potential optimization opportunities.
+This is similar to limitations placed on 'cdata' by luajit.
+Having said that, this RFC doesn't make a promise of a particular implementation making those optimizations, so should be viewed as a buffer object usability improvement first.
+
+Any chained metatables on table fields are not modified.
+This provides an option to have a more dynamic behavior, but giving up on potential performance improvements of a chained indexing access.
+This matches behavior of 'cdata' in luajit and will provide a familiarity for developers switching over.
+
+VM metatable lookups and `getmetatable` will look for the buffer object metatable first and then fall back to the global buffer metatable.
+This preserves the existing buffer extensibility point for hosts.
+
+`setmetatable` will still not be supported on buffer objects, the metatable reference is immutable.
+
+Equality checks in the VM will call `__eq` for buffers similar to tables and userdata.
+
+`__type` metatable key is ignored by `typeof`, just like it does for tables.
+As before, only host is allowed to define type names.
+This does not change how it behaved with a global host-provided metatable.
+
+Other metamethods (`__add`/`__len`/...) work as expected and do not change from how they work today for buffers with global host-provided metatable.
+
+Buffer structure size increase by the metatable aligns the data to a 16 byte boundary making it possible to store data which requires 16 byte alignment by the embedder.
+This will make buffer data match the alignment guarantee of Luau userdata objects.
+
+In order for the typechecker to understand buffers with attached metatables, we propose extending the intersections to be allowed on buffers, similar to `extern` types:
+
+`type Obj = buffer & { x: number, y: number, z: number }`
+
+## Forwards compatibility
+
+A potential evolution of the buffer type might move the data from the buffer memory block to a separate location with a redirection.
+
+Doing that will enable extending the buffer size without a major impact on performance (shrinking is a problem for range check eliminations).
+It might also allow buffer views and slices.
+
+In both cases, this change is compatible. The slice will reuse a section of the data but being a buffer it will have its own metatable.
+
+This makes it possible to have a large memory buffer and sub-allocate buffer objects with attached custom behaviors.
+
+## Drawbacks
+
+This increases buffer size by 8 bytes, with the bigger impact on 0-8 byte buffers going from 16 to 24 bytes, with an overhead decreasing with larger sizes as before.
```

Comment:
If each `buffer` needs a pointer to their own metatable, which is 1 word, and you need to know which `buffer` is host or user defined, then you need that 1 additional bit (and 7 bits padding) to store somewhere. And what happens if you need `lua_newbufferdtor` a la `lua_newuserdatadtor` (this feels inevitable)? That would incur yet another 1 word. As I understand this, it doesn't look like you can fit all of this in 8 bytes?

Reply:
There is no need to separate host and user-defined buffers, implementation of the RFC doesn't have this extra bit.

As a side note, there are 8 bits available in the buffer padding today.